src/apps/external_executor/external_executor_indexing.cpp

```diff
@@ -0,0 +1,194 @@
+  {
+    bool complete = true;
+
+    while (true)
+    {
+      std::lock_guard<ccf::pal::Mutex> guard(results_access);
+
+      const auto old_it = results_in_progress.find(key);
+      if (old_it != results_in_progress.end())
+      {
+        auto& bucket_value = old_it->second;
+
+        // We were already trying to fetch this. If it's finished fetching,
+        // parse and store the result
+        if (bucket_value.first != nullptr)
+        {
+          const auto fetch_result = bucket_value.first->fetch_result;
+          switch (fetch_result)
+          {
+            case (ccf::indexing::FetchResult::Fetching):
+            {
+              complete = false;
+              break;
+            }
+            case (ccf::indexing::FetchResult::Loaded):
+            {
+              bool corrupt = false;
+              std::string val(
+                bucket_value.first->contents.begin(),
+                bucket_value.first->contents.end());
+              bucket_value.second = val;
+              if (!corrupt)
+              {
+                bucket_value.first = nullptr;
+                break;
+              }
+              else
+              {
+                // Deliberately fall through to the case below. If this can't
+                // deserialise the value, consider the file corrupted
+                LOG_FAIL_FMT("Deserialisation failed");
+              }
+            }
+            case (ccf::indexing::FetchResult::NotFound):
+            case (ccf::indexing::FetchResult::Corrupt):
+            {
+              complete = false;
+              const auto problem =
+                fetch_result == ccf::indexing::FetchResult::NotFound ?
+                "missing" :
+                "corrupt";
+              LOG_FAIL_FMT(
+                "A file that indexer requires is {}. Re-indexing.", problem);
+              LOG_DEBUG_FMT(
+                "The {} file is {}", problem, bucket_value.first->key);
+
+              results_in_progress.clear();
+              indexed_data.clear();
+              return std::nullopt;
+              break;
+            }
+          }
+        }
+      }
+      else
+      {
+        // We're not currently fetching this. First check if it's in our
+        // current results
+        const auto current_it = indexed_data.find(key);
+        if (current_it.has_value())
+        {
+          if (complete)
+          {
+            return current_it.value();
+          }
+        }
+        else
+        {
+          // Begin fetching this bucket from disk
+          std::string blob_name = get_blob_name(map_name, key);
+          auto fetch_handle = lfs_access->fetch(blob_name);
+          std::string value;
+          results_in_progress[key] = std::make_pair(fetch_handle, value);
+          complete = false;
+        }
+      }
+    }
+  }
+
+  void MapIndex::store_data(std::string& key, std::string& value)
+  {
+    indexed_data.insert(key, std::move(value));
+  }
```

Comment:
These shouldn't be mutable refs `&`, they should either be const-ref (`const &`) to avoid copying in, or rvalue-ref (`&&`) if we're looking to move them into the LRU and can justify taking moves at the calling sites.